### DISCUSSION TITLE
OP Stack Custom Gas Token Support

### DISCUSSION THREAD

**Discussion by tynes** [wevm/viem#2197] [Idea / Feature Request]
The OP Stack is getting a feature that enables it to use a custom gas token. The spec can be found [here]. We have a guide for chain operators that can be found [here](https://oplabs.notion.site/EXTERNAL-How-to-run-a-Custom-Gas-Token-chain-d88c68306f934790be051f87213a0b1d) on how to deploy a new chain using custom gas token.

The big difference is in the deposit logic itself. A standard `approve`/`transferFrom` flow is used to deposit an ERC20 token into L2 where a native asset is minted. This allows the chain to use this custom gas token as if it was ether. It would be amazing if there was support for this in viem.

The scope of work:
- Support for calling `depositERC20Transaction` (new function on `OptimismPortal`)
- Support for calling the `IGasToken` interface on the `L1Block` contract and the `SystemConfig` contract to know if a chain is using a custom gas token and which token is the custom gas token
- Support for calling `approve` on the L1 ERC20 token that is used as the custom gas token so that it can easily be deposited into L2

**Comment by nidz-the-fact** [comment]
Hi sir @tynes , This problem is like I found an error trying to read "balanceOff".
But I deployed it as CustomGas ERC-20 <-> Native. 

![image](https://github.com/wevm/viem/assets/164603546/1f1f7581-4f89-41f9-ba96-3e813e6aa038)
![image](https://github.com/wevm/viem/assets/164603546/8cd091ec-f24d-41c2-a9d0-6a8ce0edb136)

For OP Stack CustomGas ERC20 <-> Native L2 deployments

![image](https://github.com/wevm/viem/assets/164603546/5daa7e5b-7893-46d9-9780-f03fb09648b7)
